### PR TITLE
Add check for Microsoft Windows Subsystem for Linux

### PR DESCRIPTION
**Pull request by jaredd**
These changes resolve issue #158 .  The current master branch had already solved libiio hanging due to receive errors (bug 1 mentioned in the referenced issue) while this PR handles disabling the Linux specific code for WSL which seems to have a problem with the MSG_TRUNC flag.

### CODE REVIEW

**Comment by commodo** [comment]
Change looks good.

Can you add a description in the commit comment ?
And a `Signed-off-by` tag ?

**Comment by pcercuei** [comment]
I don't think the CMake part is necesary, if running on WSL the __WSL__ macro is already automatically defined, this seems redundant.

**Comment by lclausen-adi** [comment]
I wonder do we need a way to detect this at runtime? Can native Linux binaries build on Linux run under WSL?

**Comment by lclausen-adi** [comment]
I guess my question is what happens when you use MSG_TRUNC on WSL? Does it return an error or does it just result in undefined behavior?

**Comment by rgetz** [comment]
My understanding is that it fails/returns error code. 

Chrome is doing similar things:
https://gist.github.com/therealkenc/94e0757c80113bd89a8bc22410d308be

(Hacks in code to make up for incompatibilities). 

https://github.com/Microsoft/WSL/issues/2232

Indicates : "MSG_TRUNC is supported as far as I know. "

I dont have a WSL machine to test things on. Now that appveour supports Windows Server 2016 - we should be able to increase the build matrix to include that (I think). 
https://www.appveyor.com/docs/build-environment/

**Comment by jaredd** [comment]
@lclausen-adi From memory I believe the MSG_TRUNC was returning an error, EINVAL I believe.  
@pcercuei can you elaborate on the WSL macro being predefined?  I tried a short test program:

```c++
#include <iostream>

int main()
{
#ifdef WSL
  std::cout << "WSL defined as " << WSL << std::endl;
#else
  std::cout << "WSL NOT defined" << std::endl;
#endif
  return 0;
}
```
```
$ make test_wsl && ./test_wsl
g++     test_wsl.cc   -o test_wsl
WSL NOT defined
```

I'd be happy to just use a pre-defined macro and not mod the CMakeList.txt.

**Comment by pcercuei** [comment]
@jaredd: I meant ```__WSL__``` but github ate my underscores :p

**Comment by jaredd** [comment]
Am I using the wrong compiler?  Or maybe I have the wrong WSL build?  

$ ver
Microsoft Windows [Version 10.0.16299.248]

$ cat test_wsl.cc
```c++
#include <iostream>

int main()
{
#ifdef __WSL__
  std::cout << "WSL defined as " << __WSL__ << std::endl;
#else
  std::cout << "WSL NOT defined" << std::endl;
#endif
  return 0;
}
```
$ make test_wsl && ./test_wsl
g++     test_wsl.cc   -o test_wsl
WSL NOT defined

**Comment by pcercuei** [comment]
@jaredd: well in the issue report you linked to a patch that used that macro, so I assumed it was available...

Now I agree with Lars, we should handle this at runtime, because from what I understood WSL can install whole distributions like Ubuntu and Fedora, with pre-compiled packages.

**Comment by lclausen-adi** [comment]
Maybe they just check that buf is != NULL and if it is return NULL. @jaredd can you see what happens when you use a dummy buffer for the MSG_TRUNC call?

**Comment by lclausen-adi** [comment]
Or something like:
```
ret = network_recv(io_ctx, NULL, to_trunc, MSG_TRUNC);
if (ret == -EINVAL)
		ret = network_recv(io_ctx, dst, to_trunc, 0);

**Comment by jaredd** [comment]
You guys probably know better, but could this be something simple to check with a toy test program in cmake and set a macro?  Seems a waste to try the call again and again at runtime since if it fails once, it will always fail.

**Comment by lclausen-adi** [comment]
The problem is that we do not know if we run on WSL or not when we compile the library. You can just apt-get install standard Linux packages.

**Comment by jaredd** [comment]
I see, you guys want to be able to install binaries through apt w/o worrying about whether it's WSL or not.  Yeah, I try out the suggested fix.

**Comment by lclausen-adi** [comment]
I've setup a WSL environment here. recv() yields a EFAULT when the buffer is NULL, even though MSG_TRUNC is set. Since recv() will never return EFAULT when MSG_TRUNC is set in a compliant implementation I think we should only retry if the return value is EFAULT.

If we want to avoid having to call recv() twice each time we could also add a flag to the io_context struct to indicate that we need to use the WSL workaround.

**Comment by pcercuei** [comment]
If we pass the internal buffer instead of NULL, does it work with MSG_TRUNC? And does it discard the bytes from the input queue without copying them again?

**Comment by lclausen-adi** [comment]
On WSL it returns EINVAL if MSG_TRUNC is set.

**Comment by jaredd** [comment]
You guys already have a level of indirection with the iiod_client_ops, perhaps rather than defining explicitly you could have it set via some call that accounts for WSL being detected.  Original stays the same, the nework_read_line_wsl gets set when some WSL flag is true (determined at startup).

**Comment by lclausen-adi** [comment]
That could work. I guess calling recv() with flags = MSG_TRUNC and len = 0 on the newly connected socket should do nothing on Linux, but would allow us to detect the missing feature on WSL.

I'd rather not explicitly check for WSL (e.g. by reading /proc/version) to make sure MSG_TRUNC is used once it is available in WSL.

File: network.c [hunk 1]
```diff
@@ -1389,6 +1449,18 @@ static const struct iiod_client_ops network_iiod_client_ops = {
 	.read_line = network_read_line,
 };
 
+static const struct iiod_client_ops network_iiod_client_ops_no_trunc = {
+	.write = network_write_data,
+	.read = network_read_data,
+	.read_line = network_read_line_no_trunc,
+};
+
+bool msg_trunc_supported(struct iio_network_io_context *io_ctx)
```
Comment (lclausen-adi) [inline network.c]: Should be static. The function is not used outside this compilation unit.

File: network.c [hunk 2]
```diff
@@ -1470,8 +1542,17 @@ struct iio_context * network_create_context(const char *host)
...
 	}
 
-	pdata->iiod_client = iiod_client_new(pdata, pdata->lock,
-			&network_iiod_client_ops);
+	if (msg_trunc_supported (&pdata->io_ctx)) {
+		DEBUG("MSG_TRUNC is supported\n");
+		pdata->iiod_client = iiod_client_new(pdata, pdata->lock,
+					&network_iiod_client_ops);
+	}
+	else {
```
Comment (lclausen-adi) [inline network.c]: else should be on the same line as }

File: network.c [hunk 3]
```diff
@@ -1389,6 +1449,18 @@ static const struct iiod_client_ops network_iiod_client_ops = {
...
 
+static const struct iiod_client_ops network_iiod_client_ops_no_trunc = {
+	.write = network_write_data,
+	.read = network_read_data,
+	.read_line = network_read_line_no_trunc,
+};
+
+bool msg_trunc_supported(struct iio_network_io_context *io_ctx)
+{
+	int ret = network_recv(io_ctx, NULL, 0, MSG_TRUNC);
```
Comment (lclausen-adi) [inline network.c]: While MSG_TRUNC is defined on other platforms I think only Linux supports MSG_TRUNC as a flag to recv(). Maye we should add a 
```
#ifdef __linux__
  ...
#elif
  return false;
#endif
```
around this.

File: network.c [hunk 4]
```diff
@@ -1323,6 +1323,69 @@ static ssize_t network_read_line(struct iio_context_pdata *pdata,
...
+	if (!found || i == len - 1)
+		return -EIO;
+
+	return (ssize_t) i + 1;
+#endif
+}
+
+/* use for systems that don't support MSG_TRUNC (e.g. Windows Subsystem for Linux) */
+static ssize_t network_read_line_no_trunc (struct iio_context_pdata *pdata,
+		void *io_data, char *dst, size_t len)
```
Comment (lclausen-adi) [inline network.c]: This is a bit too much copy&paste for my taste. If we update this code in the future we have to update the same code in two places. Maybe you can factor this out into a common function that takes a extra do_trunc boolean flag that is set to either true or false. And then let the compiler take care of optimizing this as it seems fit.

**Comment by pcercuei** [comment]
Could you try to do it without network_read_line_no_trunc()? Because that's a lot of duplicated code...

You could for instance store a ```bool trunc_supported``` in the context_pdata structure, initialized to true. Change it to false if the read() fails.

**Comment by jaredd** [comment]
I'll take a look.

**Comment by jaredd** [comment]
Any comments on the latest update?  I've tested this on both WSL and a OpenSUSE linux box.  Seems to work.

**Comment by lclausen-adi** [comment]
Looks good to me.

**Comment by pcercuei** [comment]
Please squash all the commits :)

From what I see the code still does the recv with MSG_PEEK. Is this working? If it is then the data is read twice (in the case MSG_TRUNC isn't supported) which is not optimal.

**Comment by lclausen-adi** [comment]
reading a large buffer twice is probably still faster than the alternative of reading the data char by char.

**Comment by pcercuei** [comment]
Yes, and reading a large buffer once is even faster :)

You could just skip the first read if trunc_supported is false, I guess.

**Comment by pcercuei** [comment]
Also, now that I think of it, this cannot work. This function is used to read a line and not more than that. If you read too much data after the CR character, the data that comes after will be lost. That will completely disrupt the communication with the server. That's why the windows code does read char-by-char, for instance.

**Comment by pcercuei** [comment]
(So that it's clear, my "this cannot work" was for my idea of a single buffer. With MSG_PEEK we can detect the CR char and know exactly how many bytes to read on the second recv call)

**Comment by lclausen-adi** [comment]
@jaredd OK, rebase, squash and then we are good to go.

Thanks for working on this.

**Comment by lclausen-adi** [comment]
Hi @jaredd,

I've tested your changes and notices two issues.
* MSG_TRUNC is not defined on Windows (at least not with the includes that we use), so this doesn't build on native Windows anymore. If fixed this by adding a #ifdef around the function.
* The msg_trunc_supported() adds a initial startup delay since the recv function waits for the timeout. I've fixed this by adding MSG_DONTWAIT.

I've pushed the fixed commit to the https://github.com/analogdevicesinc/libiio/tree/wsl-workaround branch. Let me know if you are OK with those changes and whether things still work properly under WSL.

Thanks.

**Comment by jaredd** [comment]
Changes seem reasonable.  I've tested the wsl-workaround branch on both my wsl install and a Linux box and it seems to work.

**Comment by lclausen-adi** [comment]
Closing since #171 got merged

**Pull request closed without merging**